composition-js/src/merging/merge.ts

```diff
@@ -2465,24 +2470,39 @@ class Merger {
       if (differentApplications.length === 1) {
         dest.applyDirective(name, differentApplications[0].arguments(false));
       } else {
-        const idx = indexOfMax(counts);
-        // We apply the directive to the destination first, we allows `reportMismatchHint` to find which application is used in
-        // the supergraph.
-        dest.applyDirective(name, differentApplications[idx].arguments(false));
-        this.mismatchReporter.reportMismatchHint({
-          code: HINTS.INCONSISTENT_NON_REPEATABLE_DIRECTIVE_ARGUMENTS,
-          message: `Non-repeatable directive @${name} is applied to "${(dest as any)['coordinate'] ?? dest}" in multiple subgraphs but with incompatible arguments. `,
-          supergraphElement: dest,
-          subgraphElements: sources,
-          elementToString: (elt) => {
-            const args = elt.appliedDirectivesOf(name).pop()?.arguments();
-            return args === undefined
-              ? undefined
-              : Object.values(args).length === 0 ? 'no arguments' : (`arguments ${valueToString(args)}`);
-          },
-          supergraphElementPrinter: (application, subgraphs) => `The supergraph will use ${application} (from ${subgraphs}), but found `,
-          otherElementsPrinter: (application, subgraphs) => `${application} in ${subgraphs}`,
-        });
+        const info = this.mergedFederationDirectiveInSupergraph.get(name);
+        if (info && info.argumentsMerger) {
+          const mergedArguments = Object.create(null);
+          const applicationsArguments = differentApplications.map((a) => a.arguments(true));
+          for (const argDef of info.definition.arguments()) {
+            const values = applicationsArguments.map((args) => args[argDef.name]);
+            mergedArguments[argDef.name] = info.argumentsMerger.merge(argDef.name, values);
+          }
+          dest.applyDirective(name, mergedArguments);
+          this.mismatchReporter.pushHint(new CompositionHint(
```

Comment:
Do we need this hint? Seems like customers may want to turn this on/off on a per directive basis.

Reply:
I believe it was @lennyburdette which mentioned somewhere that some users strongly feel that discrepancies between subgraphs on non-repeatable directives should error out, and I do see the argument, so this is meant as a hook to allow this.

I would agree that it's high time we allow configuring a set of hints to turn into hard errors, and that until we have that native a user would have to catch hints manually, but filtering hints manually is not that hard either and this at least allow that.

And I do get the concern for noisiness, but hints are already super noisy if unfiltered and I don't think this one is worst than others we have (but I do think this hint has more concrete usefulness than many other we have).